kv/store.go

```diff
@@ -33,55 +45,102 @@ func (store Store) CreateConnection(conn *connection.Connection) (*connection.Co
+	if err != nil {
+		return nil, err
+	}
+	ops := append(createJobs, createConnection)
+	_, err = store.client.Txn(context.TODO()).Then(ops...).Commit()
 	if err != nil {
 		return nil, err
 	}
 
-	store.Log.Debugw("Connection created.", "space", conn.Space, "connectionId", conn.ID)
+	store.log.Debugw("Connection created.", "space", conn.Space, "connectionId", conn.ID)
 
 	return conn, nil
 }
 
 // UpdateConnection udpates connection in etcd.
 func (store Store) UpdateConnection(conn *connection.Connection) (*connection.Connection, error) {
-	value, err := json.Marshal(conn)
+	_, err := json.Marshal(conn)
 	if err != nil {
 		return nil, err
 	}
 
-	resp, err := store.Client.Get(context.TODO(), string(conn.ID))
-	if resp.Count == 0 {
+	getResp, err := store.client.Get(context.TODO(), string(conn.ID))
+	if getResp.Count == 0 {
 		return nil, ErrKeyNotFound
 	}
 	if err != nil {
 		return nil, err
 	}
 
-	_, err = store.Client.Put(context.TODO(), string(conn.ID), string(value))
-	if err != nil {
-		return nil, err
-	}
-
-	store.Log.Debugw("Connection updated.", "space", conn.Space, "connectionId", conn.ID)
+	// TODO it doesn't work
+	// ops := []etcd.Op{}
+	// ops = append(ops, store.deleteJobs(conn.ID))
+	// ops = append(ops, etcd.OpPut(store.ConnectionsPrefix+string(conn.ID), string(value)))
+	// ops = append(ops, store.createJobs(conn)...)
+	// _, err = store.Client.
+	// 	Txn(context.TODO()).
+	// 	If(etcd.Compare(etcd.ModRevision(store.ConnectionsPrefix+string(conn.ID)), "=", getResp.Kvs[0].ModRevision)).
+	// 	Then(ops...).
+	// 	Commit()
+	// if err != nil {
+	// 	return nil, err
+	// }
+
+	store.log.Debugw("Connection updated.", "space", conn.Space, "connectionId", conn.ID)
 
 	return conn, nil
 }
 
 // DeleteConnection deletes connection from etcd.
 func (store Store) DeleteConnection(space string, id connection.ID) error {
-	resp, err := store.Client.Delete(context.TODO(), string(id))
-	if resp.Deleted == 0 {
+	deleteConnection := etcd.OpDelete(string(id))
+	deleteJobs := store.deleteJobs(id)
+	resp, err := store.client.Txn(context.TODO()).Then(deleteConnection, deleteJobs).Commit()
+	if resp.Responses[0].GetResponseDeleteRange().Deleted == 0 {
 		return ErrKeyNotFound
 	}
 	if err != nil {
 		return err
 	}
 
-	store.Log.Debugw("Connection deleted.", "space", space, "connectionId", string(id))
+	store.log.Debugw("Connection deleted.", "space", space, "connectionId", string(id))
 
 	return nil
 }
 
+func (store Store) createJobs(conn *connection.Connection) ([]etcd.Op, error) {
+	ops := []etcd.Op{}
+
+	numWorkersLeft := conn.Source.NumberOfWorkers()
+	numBuckets := int(math.Ceil(float64(conn.Source.NumberOfWorkers()) / float64(store.jobsBucketSize)))
+	for i := 0; i < numBuckets; i++ {
+		job := &connection.Job{
+			ID:              connection.NewJobID(conn.ID, uint(i)),
+			Connection:      conn,
+			BucketSize:      store.jobsBucketSize,
+			NumberOfWorkers: uint(math.Min(float64(store.jobsBucketSize), float64(numWorkersLeft))),
```

Comment:
Good point on the shard breakup, this is precisely where we'd need to add them for each `Job` that gets created.